src/sail_json_backend/json.ml

```diff
@@ -398,14 +398,115 @@ let remove_identity_funcs op =
+    extracted to get the default operands.
+
+    Example mapping:
+    ```
+    mapping maybe_ta_flag : string <-> bits(1) = {
+      ""           <-> 0b0, (* tu by default *)
+      sep() ^ "ta" <-> 0b1,
+      sep() ^ "tu" <-> 0b0
+    }
+    ```
+    Sample hastable representation can be seen by printing the mappings table.
+    ```
+    maybe_ta_flag: sep,tu <-> 0b0
+    maybe_ta_flag: sep,ta <-> 0b1
+    ```
+    In the above case, "tu" is extracted.
+
+    TODO: Come up with a more robust implementation to parse the operands.
+*)
+let default_operand optional optional_operand =
+  let values = Hashtbl.find_all mappings optional_operand in
+  let zero_bit_pattern s = Str.string_match (Str.regexp "0b0+$") s 0 in
+  let find_operand (l, r) =
+    match l, r with
+    | [], _ | _, [] -> None
+    | last_l :: _, last_r :: _ when zero_bit_pattern last_r ->
+        let default_value = List.hd (List.rev l) in
+        debug_print ("Found default operand: " ^ default_value);
+        Some default_value
+    | _ -> None
+  in
+  match optional with
+  | Some true -> List.find_map find_operand values
+  | _ -> None
+
 let json_of_operand k op =
   debug_print ("json_of_operand " ^ k ^ ":" ^ op);
   let opmap = List.combine (Hashtbl.find operands k) (Hashtbl.find sigs k) in
-    let opplus = remove_identity_funcs op in
-      let opname = List.hd (String.split_on_char ',' opplus) in
-        match List.find_opt (fun (name, t) -> String.equal name opname) opmap with
-            Some (name, t) -> "{ \"name\": \"" ^ name ^ "\", \"type\": \"" ^ t ^ "\" }"
-          | None -> "{ \"name\": \"" ^ opname ^ "\", \"type\": \"unknown\" }"
+      let opplus = remove_identity_funcs op in
+          let opname = List.hd (String.split_on_char ',' opplus) in
+              let optional, optional_operand = optional_operand k op in
+                  let default = default_operand optional optional_operand in
+                      let optional_str = match optional with
+                        | Some b ->  string_of_bool b
+                        | None -> "false" in
+
+                      match List.find_opt (fun (name, t) -> String.equal name opname) opmap with
+                      | Some (name, t) ->
+                          "{ \"name\": \"" ^ name ^ "\", \"type\": \"" ^ t ^ "\", \"optional\": "
+                          ^ optional_str
+                          ^ (match default with
+                            | Some d -> ", \"default\": \"" ^ d ^ "\""
+                            | None -> "")
+                          ^ " }"
+                      | None ->
+                          "{ \"name\": \"" ^ opname ^ "\", \"type\": \"unknown\", \"optional\": "
+                          ^ optional_str
+                          ^ (match default with
+                            | Some d -> ", \"default\": \"" ^ d ^ "\""
+                            | None -> "")
+                          ^ " }"
```

Comment:
There is a fair amount of duplicated code here. Could you refactor this?

Reply:
Let me know if the refactored one smells right. I have tried to factor out the chunkier redundant part.